File: 0139-word-break/0139-word-break.ts
function wordBreak(s: string, wordDict: string[]): boolean {
    let memo: boolean[] = new Array(s.length).fill(undefined);
    return rec(0, s, wordDict, memo);
}

function rec(idx: number, s: string, wordDict: string[], memo: boolean[]): boolean {
    if (idx === s.length) return true; // If we've processed the entire string, return true
    if (memo[idx] !== undefined) return memo[idx]; // Check memoization

    for (let i = idx; i < s.length; i++) {
        let prefix = s.substring(idx, i + 1);
        if (wordDict.includes(prefix) && rec(i + 1, s, wordDict, memo)) {
            return (memo[idx] = true);
        }
    }

    return (memo[idx] = false);
}

File: 0139-word-break/0139-word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
       Set<String> st = new HashSet<>(wordDict);
       int maxlen = 0;
       for (String word : wordDict){
       maxlen = Math.max(word.length(),maxlen);
       } 
        boolean[] dp = new boolean[s.length()+1];
        dp[0] = true;

        for (int i = 1; i <= s.length(); i++){
            for (int j = i-1; j >= Math.max(0,i-maxlen); j--){
                if (dp[j]&&st.contains(s.substring(j,i))){
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[s.length()];
    }
}